### ISSUE TITLE
EarlyStopping for distributed training job

### ISSUE THREAD

**Issue by g-karthik** [pytorch/ignite#664]
https://github.com/pytorch/ignite/blob/87a42e3c5336317f2582262fa08cab24df36b0a0/ignite/handlers/early_stopping.py#L6

What is the expected/observed behavior of this class when trying to "early stop" a distributed training job?

Secondly, how would one go about implementing this behavior? I'm currently all-reducing the negative log-likelihood and have it as an additional ignite metric attached to my evaluator. Should my scoring function return this all-reduced metric or should it return the original metric?

I would imagine early stopping ought to occur at a GPU-level, rather than at a node-level. Could you please shed some light on this?

@vfdev-5

**Comment by vfdev-5** [comment]
@g-karthik thanks for the feedback. That's true that early stopping is not unit-tested in ignite in distributed configuration. Let me take care of it.

The behaviour I would expect of EarlyStopping handler in distrib config is that the trainer on all processes (ranks) have this handler (not only rank == 0) and while being attached to an evaluator it should have access to all-reduced metric value through the score function. Idea is that all ranks has the same logic inside EarlyStopping and if they need to stop the trainer, they should stop all trainers on all processes. What do you think ?

```python
def get_default_score_fn(metric_name):

    def wrapper(engine):
        # here score should be all-reduced
        score = engine.state.metrics[metric_name]
        return score

    return wrapper

score_function = getattr(config, "score_function", get_default_score_fn(metric_name))
es_handler = EarlyStopping(patience=config.es_patience, score_function=score_function, trainer=trainer)
evaluator.add_event_handler(Events.COMPLETED, es_handler)
```

**Comment by vfdev-5** [comment]
I'll keep it open for time of discussions

**Comment by g-karthik** [comment]
@vfdev-5 thanks for the quick reply and for adding the tests! Indeed, I think we should early-stop using the all-reduced metric.

But if we were to use the process-level metric, we'd end up early-stopping each process separately (at different points in time, presumably), leading to a different kind of early-stopping that depends on the actual batches sent to each GPU. I wonder if such an early-stopping makes sense.

Also, I have a secondary unrelated question: is it possible to have multiple early stopping handlers associated with multiple evaluators for a given trainer? It would be useful if there are multiple validation sets to work with. I'm thinking of using differing patience values for them.

**Comment by vfdev-5** [comment]
> But if we were to use the process-level metric, we'd end up early-stopping each process separately (at different points in time, presumably), leading to a different kind of early-stopping that depends on the actual batches sent to each GPU. I wonder if such an early-stopping makes sense.

@g-karthik I think if we would stop one process-level trainer by such process-level early-stopping, other trainers still then continue the training and probably would freeze on the next grads reduction op when we update the model. Probably, this is not what we would like to have... 


> Also, I have a secondary unrelated question: is it possible to have multiple early stopping handlers associated with multiple evaluators for a given trainer? It would be useful if there are multiple validation sets to work with. I'm thinking of using differing patience values for them.

I would say, yes. IMO, in this case you will need to define N evaluators for N validation datasets and to each evaluator we can add one earlystopping. All earlystopping handler can stop the only one trainer. So, the condition to stop the training is roughly if one of earlystopping conditions is satisfied.

**Comment by g-karthik** [comment]
> 
> @g-karthik I think if we would stop one process-level trainer by such process-level early-stopping, other trainers still then continue the training and probably would freeze on the next grads reduction op when we update the model. Probably, this is not what we would like to have...
> 

Yeah that makes sense! Also we wouldn't want the main rank-0 process to early-stop and die this way, while the others are running.

> 
> I would say, yes. IMO, in this case you will need to define N evaluators for N validation datasets and to each evaluator we can add one earlystopping. All earlystopping handler can stop the only one trainer. So, the condition to stop the training is roughly if one of earlystopping conditions is satisfied.

Yep, exactly, that's the kind of behavior I'm looking for, thanks!

**Comment by vantainguyen** [comment]
I think we can apply earlystopping there with some patience. When one stops other may still run and if the weights are still good to boost the performance of the model, then that's good. Otherwise, we already have the good ones saved by modelcheckpoint. No harm caused.
